### DISCUSSION TITLE
New modular codebase

### DISCUSSION THREAD

**Discussion by DarkenLM** [aclist/kbin-megamod#216] [Active discussions]
Hello. I've recently decided to create a new mod for KES, and I found the workflow to be tiresome and not ideal, in my opinion. So, I created my own workflow for development, with a POC available here: https://github.com/DarkenLM/kbin-kes/tree/rewrite

I added to the POC features I found to be useful, such as:
- A local development server to completely bypass github, removing delays due to github cache refreshes.
- Easy to use build system: The command `pnpm run dev` is all that is required to create a development-mode version of the userscript and automatically spins up a local development server, and the command `pnpm run build` creates a production-mode version of the userscript, ready to be pushed to github.
- Modular development: Instead of having of having to modify the manifest.json file and the mod source on different places, each mod now lives on it's own directory inside the src directory, along with it's own manifest.json file, and are automatically processed by the build system. It also allows for individual modules to be excluded from the build by using the `-e [module name]` flag on both build scripts.
- Typescript support: I prefer to use Typescript due to type safety, but Javascript is also accepted by the build system with no further configuration required (Examples of Typescript mods can be found [here](https://github.com/DarkenLM/kbin-kes/tree/rewrite/src/correct_user_comments) and [here](https://github.com/DarkenLM/kbin-kes/tree/rewrite/src/comment_highlighting)).

The POC also allows for easier contributions, as the only changes that need to be sent are contained in single directory inside the src directory, possibly avoiding merge conflicts on the main manifest.json.

To generate the KES userscript, I converted the [gen.sh](https://github.com/aclist/kbin-kes/blob/main/gen.sh) file, and adapted it to the rest of the build system. I did not convert any of the other scripts.

I was wondering if my POC would be acceptable to be integrated into the KES project.

**Comment by aclist** [comment]
Hello, thanks for your message.

I share your frustrations with GitHub's caching. However, there are some misconceptions here regarding how scripts are supposed to be debugged and integrated. I'll try to answer in as comprehensive a manner as possible and maybe we can find a method that incorporates some of your suggestions.

The idea of making manifest artifacts atomic is intriguing.

1a. Regarding debugging: GM supports arbitrary URI schemas in the `@requires`. These can be any remote URL or even a local filepath. Therefore, there is no need to rely on GitHub when testing scripts locally. It is enough to simply use a local file, a gist, or some other remote URL updated ad-hoc to see changes immediately.
1b. In the standard workflow for mod submission, a mod author creates a single atomic mod and tests it within the context of kbin. They do not need to integration test it within KES. The mod will necessarily work when sourced into KES. It is the responsibility of the project maintainer to use the `gen.sh` and other build tools to prepare the master `kes.user.js` and integration test mods that come in via PRs. The only responsibility of the mod author is to verify that the GM script works by itself in a browser. Perhaps the wording about 1a and 1b is not sufficiently clear in the documentation?
2. The KES maintainers want the KES build process to be as "vanilla" as possible and eschew nodejs, npm, and other fancy frontend web development systems. Insofar as GM scripts are simply generic JS code run locally in a browser, and KES is an agglomeration of such code, the cost/benefit tradeoff from requiring heavy upfront dependencies seems minimal given that what needs to be achieved here can be done without introducing complex systems.
3. Another stated goal of KES is reducing the barrier to entry for non-technical users who may dabble in script creation but are not accustomed to highly engineered build systems or web frameworks. Your solution seems to support both TS and JS, so this is not an issue per se, but I question whether this build system itself would raise the barrier to entry for inexperienced users, who make up the majority of mod authors.
4. Speaking as one KES maintainer, personally, I would not use such a tool since I find npm tedious to use and don't want to install extra dependencies for the sake of something simple. The `gen.sh` is intentionally a naive shell script because it "just works." If you find this new system to be idiomatically more intuitive for you, I think it is fine if we agree on a shared specification for how to structure mod directories so that they work in either system. 

To summarize, in the typical workflow:

- An author creates and locally tests a script in GM using their browser and a local file
- An author submits the file and some manifest markup as a PR
- A KES maintainer reviews the mod and integration tests it
- A KES maintainer generates the new version of KES, which incorporates said mod

Are there scenarios where an author would need to integration test a mod within KES? If so, could you enumerate such scenarios? Presumably these might involve wanting to see how the configuration settings will look in-place? There is a very rudimentary, ugly POC located under `/manifest_gen` which is intended to be a "visual JSON generator" that would let the mod author fill in a set of fields and see what the KES menu would look like with them. It generates the desired JSON output and copies it to the clipboard. The visual component is merely a placeholder and has not been implemented. If I recall correctly, all of the form fields and permutations thereof work correctly, and although basic, this form allows an author to copy presanitized JSON to the clipboard and simply paste it in at the end of the bigger manifest.

I think the idea of having atomic manifest chunks that get ingested into the master manifest by the KES generator is attractive, as users could just ship a single JSON object bundled with their PR and not have to worry about whether it is nested correctly in the master manifest.

I am wondering about how we can consolidate the your approach with this approach without adding layers of complexity. Could we update the `gen.sh`, to be run by a maintainer, to ingest these JSON shards into a single file? The one-mod, one-directory structure doesn't seem problematic. Alternatively, you could have a directory of mods and a directory solely containing manifest chunks, with the filename prefix being identical to the corresponding mod name. I think the latter might be better since it somewhat simplifies the path to these files:

`slug/mods/modname.js` and `slug/manifests/modname.json`

VS

`slug/mods/modname/modname.js` and `slug/mods/modname/modname.json`

This is merely a semantic issue, but it seems slightly less verbose and less nested IMO. I understand that having mods within their own directory makes them inseparable from their manifest, but I don't necessarily think this is a problem, since the manifest only works within the context of KES anyway.

Being able to whitelist/blacklist certain mods from the build for testing purposes could be helpful, but is a CLI for this necessary? It seems like it would lead to very long concatenations of CLI flags. Perhaps a simple flat include/exclude file that is used to tell the generator to skip mods?

I hope this does not seem unnecessarily negative, but I question whether this level of abstraction is necessary for what is ultimately a glorified GM script. Could you go into a bit more detail about what aspects of mod creation are tedious or confusing right now?

**Comment by DarkenLM** [comment]
Thank you for the answer. I'll try to answer each point individually.

1a. In my POC, I favored integrated testing over isolated testing, as a means to more quickly resolve conflicts with other mods, whether on the DOM or the code of the userscript itself. I added the development server as `file://` URIs seemed to be very unreliable with TamperMonkey, and sometimes not working at all.
1b. When developing one of the new mods, I found an incompatibility with an already existing mod (omni, IIRC), with both mods modifying an element leading to the last one to be initialized to fail. I figured it could happen with other mods as well, hence building the entire script during development, to eliminate those possible incompatibilities.
2. I introduced node.js (and by extension, npm, though it is mostly used for compiling the typescript files into javascript) to make it cross-platform, as the POC builds the userscript during development. The heaviest dependencies in there are esbuild and typescript, used to compile the typescript files, and pnpm, though that one can be easily removed by using npm directly. These three packages make about half the size of the node_modules, and other three can be `vendor`ed, though it would increase the size of the project itself. I could modify the build system to not compile typescript by default and make it a plugin, removing the heavy dependencies from the project.
3. I don't think the entry barrier would be high with this build system. The only commands needed to make the project work are `pnpm install` and `pnpm run dev`, and eventually `pnpm run dev -e [module]`. Compared to build systems like Vite, which require heavy configuration, I think this system is extremely simple. If the entry barrier is still too high, twin batch/shell files can be made to make the tasks easier, avoiding opening the terminal entirely.
4. I created this build system mainly for simplifying (from my POV) some parts that I found frustrating during development, and cross-platform compatibility was one of them. I understand why you wouldn't use a tool like this, and I can also see installing node.js for a non-experienced users. But I am open to discuss a shared specification.

I saw the `manifest_gen` directory, and tried to use it, but I preferred, as you guessed, to tweak the mods' manifest manually using the master manifest.json as a reference and seeing the settings on KES. As for the need to integration test individual mods, refer to 1b.

I prefer the `slug/mods/modname/modname.js`/`slug/mods/modname/modname.json` structure mainly because when there are dozens, and eventually hundreds of mods, it would become annoying to navigate between the two directories and find the corresponding file for each mod if they were separated, compared to having all files for the mod together on one directory.

I created the CLI as a means to having the entire build system available from the terminal during development. I agree excluding various mods would lead to very long commands, and I have thought about the issue myself. Currently, the build system uses two files for configuration:
- `.build/config.build.cjs`, which contains all internal settings for the build system.
- `tsconfig.json`, which currently only contains the configuration for ephemeral types (imports like `import { genericXMLRequest } from "@kes/safeGM";` allow typing the KES library functions on TS files, but are removed from the output files).

I don't think an includes/excludes configuration should be placed within the internal config, as it would confuse the regular users, but the tsconfig can also lead to confusion, as an `exclude` property is already present within, which serves a completely different purpose. Perhaps creating a new configuration file at the root could work, but I am still unsure what to name it.

I do not see your comment as unnecessarily negative at all. As for the abstraction this system provides, it can be seen as unnecessary, but I think it would be helpful to keep a layer of abstraction that allows the underlying code for KES to be changed without necessarily needing to change all mods. Currently, KES is a GM script, but it could evolve to something like an extension, like RES once was, and having a build system like this one could reduce the work required to migrate.

Finally, summarizing what I see as confusing and tedious on mod creation with the current workflow:
- Testing compatibility between different mods is hard.
- Testing the mod within the context of KES is confusing. I found a sneaky bug that occurred due to KES initializing the mods in a sequence I was not expecting, which completely broke my mod.
- It's annoying to scroll hundreds of lines on the manifest to reach the mod's specific manifest.

There are other small annoyances, but they're closely related to the ones I listed, unless I forgot some.

**Comment by aclist** [comment]
You raise some valid and interesting points, particularly the issue of inter-mod compatibility with respect to loading of elements. This is not dissimilar to an issue called "load order" in the context of mod managers, where the serial order of mods has to be tweaked to give specific ones precedence so that they don't attempt to edit a resource that was already edited by a prior mod. Thus far this happens infrequently with KES, but you rightly note that there may be some mods where a resource has already been edited, or a race condition applies. Perhaps we need to be very specific about the load order here when sourcing scripts, rather than resorting to alpha order.

It is true that the shell script is not (strictly speaking) cross-platform, though this reveals my own biases for a preference towards development on Linux systems. I think it is fine if we incorporate the node build system provided both systems could coexist as an optional routine for maintainers and/or mod authors who want to integration test a priori.

Perhaps `.modinclude` or `.modexclude` would suffice as a simple flat file with the name of the mods.

In theory, keeping the JSON hunks atomic would still function with the existing `manifest_gen` for those users who want a GUI-driven approach to generating the JSON snippet and don't want to/can't understand the JSON syntax. Since the JSON emitted is a complete object, they could just paste it in to the file in their mod's unique directory and wouldn't need to worry about fitting it into a larger manifest and breaking the nesting, which has happened from time to time. In that sense, I think this could be the start of a basic shared specification. I am amenable to moving to a one-directory-per-mod structure and willing to refactor the shell script to follow this spec.

Your points about future expansibility are well taken, as there have been some users, particularly those in a mobile context, who are looking for something more plug-and-play that is not constrained by the need to manually install a script into a Greasemonkey-type manager. It could definitely be interesting if you want to take an active role in structuring such a trajectory for the project and focusing on more robust code. All that being said, I'm not sure if the project timeline is currently at a point where a packaged version (web app, extension, or otherwise) is truly necessary, but it could be viable in the future if sourcing all of these remote resources at runtime starts to become untenable.

For the time being, what actionable steps would you like to take so that we can incorporate your changes and settle on a set of rules that will satisfy all parties below?

- KES maintainers
- KES contributors (technical)
- KES contributors (non-technical script authors)
- Platform-agnostic users

I have a strong preference here for not enforcing any build systems if we can avoid it; i.e., a maintainer should be able to ingest the mods, package the final file, and ship it via basic shell scripts if they so choose, without introducing additional overhead. By contrast, a (e.g.) Windows user could also use additional node dependency files if that helps them.

**Comment by DarkenLM** [comment]
One problem I've found with files like `.modinclude`. or `.modexclude` is their restrictive nature. If more configuration properties are required in the future, either the system needs more configuration files, or needs to move to a new file with all properties, though this problem can be easily signaled by bumping a new major version.

As far as I understand, the build system must be compatible with pure shell scripting, even if the shell scripts require some modifications.
To make the system compatible, a few things could be added/changed:
- I could make the build system itself modular, excluding the heavier dependencies from the base system, making it as lightweight as possible from the get-go. Typescript support could be added through a plugin that users could install.
- A shared config for both systems. Currently, making a shell script read the `config.build.cjs` file will be challenging, as it is written in javascript, and some properties require dynamic data. The config could possibly be turned into a static json file with a syntax for dynamic commands, similar to bash's `$(...)`, that could be interpreted both by the javascript system and the scripts.

If you have any other requirements, please let me know @aclist.

**Reply by aclist** [reply]
I am sorry it took a few days to reply; I was out of town this week.

I may have communicated the requirements above in a confusing way, so I'll try to restate it: to keep things as simple as possible, I think the different build systems/approaches should be independent of each other, but produce the same result. The outcome should be, so to speak, idempotent--whether you use the shell script or your build system, the generated files will be the same. Mod authors need only concern themselves with placing their files in the requisite location. Refactoring the shell script into a file for another build system should be a one-time operation.

Whether the maintainer or developer wishes to use one or other build system should be immaterial here. Similarly, whether the exclusions/inclusions are sourced via a particular flat file or via CLI parameters should not matter, since this is decided by the person invoking the build. Your system need not read .modinclude files if it does not want to, as those files are not included a priori in the codebase; it would just be a drop-in configuration used by a given builder of the files.

So mainly what I am trying to decide here is:

- What should be the output format that is uniformly produced by both approaches?
- What should be the required input files submitted by mod authors?

I think that if we start trying to make the shell scripts too intertwined with other approaches or vice versa, then while it may work in the interim, we are going to have this problem yet again if a third party creates another alternative build system. Therefore, the requirements to output a final KES file should be utterly generic.

I've also been meaning to ask you how you reconciled the problem with load order when testing your mod with the omnibar mod. Should explicit load order be something defined a priori in the build system, e.g., load mod X in position 1, followed by mod Y, and so on? Is this overkill?

To reiterate the problem, the build system need not be compatible with shell scripts; the build system need only be "compatible" insofar as it creates the same output. 

In light of the above:

> I could make the build system itself modular, excluding the heavier dependencies from the base system, making it as lightweight as possible from the get-go. Typescript support could be added through a plugin that users could install.

Strictly speaking, I am less concerned with the "heaviness" of the build system if it is decoupled from the shell script method and is not mandatory. 

> A shared config for both systems. Currently, making a shell script read the config.build.cjs file will be challenging, as it is written in javascript, and some properties require dynamic data. The config could possibly be turned into a static json file with a syntax for dynamic commands, similar to bash's $(...), that could be interpreted both by the javascript system and the scripts.

I took a look at the `config.build.cjs` -- does it seem likely that the core configuration parameters will change much in the future? That is to say, I don't think a shell script needs to read this file or vice versa. The parameters we are supplying are mostly monolithic, such as instance TLDs, author name, some grants/deps, etc. If we add a new instance or dependency here or there, this amounts to a one line change in both files, so I don't think they need to be in lock-step per se. Besides, isn't it the responsibility of a third party build system to parse or rewrite the original configs from the shell script and munge them into a new format? After bootstrapping this alternative config file once, it shouldn't be necessary to do again, other than transitory changes.

If you think this is still brittle, why not just have flat files listing the deps, instances, etc.? Then any build system could just ingest those files and be abstract.

**Comment by DarkenLM** [comment]
Sorry for a late response as well.

I might have interpreted your requirements wrong before, but I think I've got a solid idea of them now. Answering your points in order:
- My build output currently uses `dist/<mod_name>` for each mod, with it's code and manifest, which is read by the `genKES.build.cjs` script and builds the master manifest on `dist/helpers/manifest.json`. The `kes.css`, `ui.json` and `safegm.user.js` are copied from `lib/helpers` to `dist/helpers`. The KES script itself is stored on `lib/kes.js`.  I separated those to allow the KES files to be separated from the mods, allowing the build system to build the mods as required without having to mess with KES itself. I could make the build system flatten out the mod files to make it use the current tree. I believe, however, that the global development should have the KES files on their own directory, for organization.
- For my build system, only the mod directory along with it's files would be necessary, which in my opinion is way easier than the current way.
- I solved the omnibar problem by manually reordering the `@require` lines on the KES script. I have yet to implement such a feature on the build system itself, though I'd probably use either a permission override system like:
  ```
  {
    "default": 0,
    "omni": -1
  }
  ```
  and order the mods from highest to lowest. One other option would be to explicitly declare which mods should be loaded before and after the current mod, though it would probably require more work due to possible circular references.

Finally, if you'd like to use a flat config system, I'd be willing to use as well, but I'm willing to also leave the config unchanged and leave it to the responsibility of the developer of the third-party build system.

**Reply by aclist** [reply]
Collecting my thoughts below:

- Is omnibar modifying a resource it shouldn't be? I'm happy to refactor it if it is touching something it doesn't need to. Alternatively, and perhaps necessarily:
- Most mod loaders use a naive implementation for load order by indexing each mod at a given position after manual testing. Albeit monolithic, this ensures that the load order is explicit. It seems like the alpha order was mostly working with the exception of omnibar, so we could start with the known working alpha order, index those accordingly in sequence, and then move omnibar to index 1. Given that you proposed a system of subdirectories, this would function well in terms of keeping them ordered. For example:
    - 001-omnibar
    - 002-someothermod
    - 003-someothermod
- Under the proposed system, KES can ship with other third-party build systems if they are PRed and integrated, and such build systems must generate identical files in identical locations. It is the responsibility of the user to choose the desired build system.
- One issue I see with the above is that your system supports both .ts and .js, but this is not backwards compatible with, e.g., the original/primitive KES shell script insofar as a .ts file alone willnot work. To ensure a minimum of compatibility, mod submissions should be in .js format and optionally in .ts format. They can be both, but .js must be included. Does this make sense?
- For reasons of semantics, I believe user-facing GM scripts (in this case, KES) must have the extension .user.js, such that they can be auto-installed when a user points their mod manager to the URL.
- I propose the below hierarchy. The reason is because, despite having named subdirectories, having multiple files named `manifest.json` floating around inside of each could be problematic if the manifest chunks become detached from the mod at some point during a future migration:
    ```
    mods/INDEX-modname
  ├──modname.js -- the mod itself
  └──modname.json -- the manifest object for the mod
- The manifest build process should step through the subdirectories and build a JSON file consisting of an array of objects, one per mod.
- I don't fundamentally see an issue with the current tree, with KES in the root and the manifest inside helpers. I believe this is consistent with the logic of how other repositories handle userscripts.
- One important thing to note about this change is that there are, to my recollection, some user-submitted mods that actually bundle multiple mod functions into a single .js file, yet declare them as separate features within the master manifest (IIRC, hide upvotes/downvotes). I believe this was just done by the user's preference, but it would deviate from the above spec, so these would have to be split up into separate mods with their own manifests.

**Reply by DarkenLM** [reply]
In response:
- When I tested it, omnibar tried to access an undefined variable named "kt" (line 450 of the omni.user.js file), and everything after that instruction simply halted, possibly due to error bubbling.
- Given it's a simple solution and the focus of the project on simplicity, I don't see a reason not to use it. No further modifications on my end need to be made, the build system processes the modules in the order they are present on the filesystem.
- No comment on this one.
- The user can send the mod folder present on the build output, excluding the `build.manifest.json` file. Given no imports are used, no transpilation artifacts are present and can simply be used as source code.
- The kes main script present on the lib directory was named without the ".user" as I temporarily used the lib directory to generate the userscript in a previous version of the system, and it remained that way. But it can be easily changed to `kes.user.js`.
- Though it would require a bit of modification on my end, I have no problems with this hierarchy.
- That's the current implementation on my build system.
- I personally dislike including source code on the root of a project and use the root exclusively for configuration files or documentation (README, LICENSE, etc), due to some problems I had in the past with path resolutions becoming way too confusing and sacrificing modularity.
- Yes, it would require changes on those monoliths. I propose either:
  - Implementing a hard limitation, refusing to build a mod if it is a monolith.
  - Implementing a soft limitation, printing a deprecation notice to the console, but building anyway.
  - Add support for monolithic mods, which would allow for mod grouping, though it would require a new property on the mod manifest, for source files, which would need to be processed individually.

Also, where should the third-party build systems be included in the project, assuming more may be developed?

If we can define a spec for the project, I will implement it as soon as possible on the build system.

**Reply by aclist** [reply]
- Looks like the omnibar mod is pretty terse in its declarations, but found the problem and fixed it on the testing branch.
- I believe that different GM extensions implement their interpreters differently; some are strict and will necessarily fail on exceptions, while others will let them silently pass. Anyway, I think it's meaningful as a matter of SOP to test and see what collisions there might be between mods and observe possible race conditions at runtime, so I'm thinking of holding on the proposed indexed mod order as a method of last resort. It is easy to add in at a later time if necessary. As you note, it doesn't change the behavior of the build tools. Ideally, mods should be designed in such a way that they don't try to edit the same resources, but it's not out of the realm of possibility that some future mod may intentionally wish to augment the behavior of some other mod (undecided how I feel about this, though), so perhaps load order could be important in the future.
- Noted.
- Noted.
- Alright. I am going to start updating the shell script to support this approach and concatenate the JSON objects.
- I don't disagree with that on an academic basis, but the KES repo is a bit weird in that it is both source code and a user-facing script that should be easy to find without digging through directories.  The releases section is merely used as a courtesy to print atomic patch notes and checksums, but users don't actually get new releases from there. There is also the issue of existing inbound links posted through kbin et al pointing to the kes.user.js in the root path, so playing around with this hierarchy at this point potentially introduces dead links, so better to preserve compatibility here.
- It seems sane to enforce the rule of atomic mods and break up any existing mods that are concatenations of multiple mods. I guess one user decided to do that because they seemed semantically related, but it sacrifices modularity for ease of development (?).

> Also, where should the third-party build systems be included in the project, assuming more may be developed?

Open to suggestions here. It seems there should be a dedicated directory for builders/generators, and anything that is responsible for taking input files and generating output should be nested under there. This would include `gen.sh`, the manifest_gen, other build tools, etc., each under a named directory.

Incidentally, this set of changes breaks the old method of submitting mod PRs, so the version will be bumped to 3.0.0 to flag this change.

**Reply by aclist** [reply]
On the testing branch, I've gone ahead and updated the structure of the codebase per our discussion and bumped the major version to 3.0.0:

1. Mods get atomic directories with a script and JSON manifest object inside:
```
├── unblur
│   ├── unblur.json
│   └── unblur.user.js
└── user-instance-names
    ├── user-instance-names.json
    └── user-instance-names.user.js
```
2. Mods that sourced multiple mods (functions) from within the same mod have been split into atomic mods (1 manifest entry = 1 mod, no overlap)
3. The `build` directory has been newly created at the top level. Currently it contains the `scripts` subdirectory and `manifest_gui` directory. Scripts is a collection of scripts that are used for build routines, such as generating KES and generating the manifest. `manifest_gui` is one example of a build tool that could be used for generating atomic manifest objects. Your build system could go under `build` and can emit `kes.user.js` and `manifest.json` where it wants to, but it is the final responsibility of the builder/user to move those files into, respectively, the root and `helpers`. 
4. `kbin-mod-options.js` was moved into the `helpers` directory since it is not a mod, but merely a utility function that is sourced at runtime.
5. Per the above, the correct URI to mods in the requires is as follows. Notice that the mod subdirectory and mod prefix share the same name (in this example, "subs").
```
// @require      https://github.com/aclist/kbin-kes/raw/testing/mods/subs/subs.user.js
```
6. The `gen_kes.sh` build script concatenates all of the JSON objects under `mods` and puts them into an array (`manifest.json`). This script is intended to be run by the repo maintainer and triggers off of pre-commit hooks.
7. The `bef` script was added to `scripts`. This is merely a utility script that stitches two images together. This will be used to create preview images of before/after functionality for mods.

I believe that covers everything from our discussion, with the remaining TODOs:
- [ ] Update documentation instructing authors to add manifest objects into their own subdirectories only
- [ ] Advise authors to normalize the subdirectory, mod, and manifest hunk to use the same name
- [ ] Future proofing: drop GM user.js suffix and headers from mods, as they are not intended to be loaded directly into GM

I have generated KES from the testing branch and done some preliminary sanity testing. Everything appears to be working as before. Only the general location of the mods has changed, but everything else should be the same. Nonetheless, please stress test.

As of this writing, no priority ("load order") is given to specific mods. Per an earlier comment, I refactored the omnibar mod to fix the halting behavior.

**Reply by DarkenLM** [reply]
I've taken a look at the branch and I don't think it will require many changes in order to make my build system fit inside the build directory. The only thing I'll have to test is whenever tsconfig will be able to recognized by VSCode inside the build directory for intellisense purposes. If I recall correctly, it might require a tsconfig at the root, or a symlink for the mods directory inside the build system.

Also, sorry for leaving the conversation on hold for so long, I've been dealing with some personal issues during the last few months, but I'll try to implement the changes in the following weeks.

**Reply by aclist** [reply]
It is alright, I couldn't give this project much focus time until now.

It looks like there have also been some technical issues/spam plaguing kbin for the last month, so there has been an overall reduction in activity on there.
